shared/github/__init__.py

```diff
@@ -0,0 +1,86 @@
+import logging
+from datetime import datetime
+from time import time
+from typing import Optional
+
+import jwt
+import requests
+
+import shared.torngit as torngit
+from shared.config import get_config, load_file_from_path_at_config
+from shared.utils.cache import cache
+
+log = logging.getLogger(__name__)
+
+loaded_pems = None
+
+pem_paths = {
+    "github": ("github", "integration", "pem"),
+    "github_enterprise": ("github_enterprise", "integration", "pem"),
+}
+
+
+def get_pem(pem_name: str) -> str:
+    path = pem_paths[pem_name]
+    return load_file_from_path_at_config(*path)
+
+
+@cache.cache_function(ttl=480)
+def get_github_integration_token(service, integration_id=None) -> Optional[str]:
+    # https://developer.github.com/apps/building-github-apps/authenticating-with-github-apps/
+    now = int(time())
+    payload = {
+        # issued at time
+        "iat": now,
+        # JWT expiration time (max 10 minutes)
+        "exp": now + int(get_config(service, "integration", "expires", default=500)),
+        # Integration's GitHub identifier
+        "iss": get_config(service, "integration", "id"),
+    }
+    token = jwt.encode(payload, get_pem(service), algorithm="RS256")
+    if integration_id:
+        api_endpoint = (
+            torngit.Github.api_url
+            if service == "github"
+            else torngit.GithubEnterprise.get_api_url()
+        )
+        headers = {
+            "Accept": "application/vnd.github.machine-man-preview+json",
+            "User-Agent": "Codecov",
+            "Authorization": "Bearer %s" % token,
+        }
+        url = "%s/app/installations/%s/access_tokens" % (api_endpoint, integration_id)
+        res = requests.post(url, headers=headers)
+        if res.status_code in (404, 403):
+            log.warning(
+                "Integration could not be found to fetch token from or unauthorized",
+                extra=dict(
+                    git_service=service,
+                    integration_id=integration_id,
+                    api_endpoint=api_endpoint,
+                ),
+            )
+            return None
```

Comment:
Oh, this is a code change.  Code in worker raises a custom exception here.

Reply:
Hum... should we raise something here as well?

Reply:
Yeah I guess I can create a custom exception `shared` and then catch it in `worker` and re-raise the other custom exception.